File: depth-first-search/depth_first_search.php
<?php
function dfs($graph, $start, &$visited) {
    $visited[$start] = true;
    echo $start . " ";
    
    if (isset($graph[$start])) {
        foreach ($graph[$start] as $neighbor) {
            if (!isset($visited[$neighbor])) {
                dfs($graph, $neighbor, $visited);
            }
        }
    }
}

$graph = [
    0 => [1, 2],
    1 => [2],
    2 => [0, 3],
    3 => [3]
];
$visited = [];
dfs($graph, 2, $visited);
?>

File: depth-first-search/depth_first_search.rb
def dfs(graph, start, visited = Set.new)
  return if visited.include?(start)
  
  visited.add(start)
  print "#{start} "
  
  (graph[start] || []).each do |neighbor|
    dfs(graph, neighbor, visited)
  end
end

graph = {0 => [1, 2], 1 => [2], 2 => [0, 3], 3 => [3]}
dfs(graph, 2)